src/Storages/RocksDB/EmbeddedRocksDBBulkSink.h

```diff
@@ -0,0 +1,67 @@
+#pragma once
+
+#include <condition_variable>
+#include <stdatomic.h>
+#include <Processors/Sinks/SinkToStorage.h>
+#include <rocksdb/db.h>
+#include <rocksdb/status.h>
+#include <Common/CurrentThread.h>
+#include <Common/ThreadStatus.h>
+#include <Common/ThreadPool.h>
+#include <Columns/ColumnString.h>
+#include <Processors/Chunk.h>
+
+
+namespace DB
+{
+
+class StorageEmbeddedRocksDB;
+class EmbeddedRocksDBBulkSink;
+struct StorageInMemoryMetadata;
+using StorageMetadataPtr = std::shared_ptr<const StorageInMemoryMetadata>;
+
+/// Optimized for bulk importing into StorageEmbeddedRocksDB:
+/// 1. No mem-table: an SST file is built from chunk, then import to rocksdb
+/// 2. Overlap compute and IO: one thread prepare rocksdb data from chunk, and another thread to write the data to SST file
+class EmbeddedRocksDBBulkSink : public SinkToStorage, public WithContext
+{
+public:
+    EmbeddedRocksDBBulkSink(
+        ContextPtr context_,
+        StorageEmbeddedRocksDB & storage_,
+        const StorageMetadataPtr & metadata_snapshot_);
+
+    ~EmbeddedRocksDBBulkSink() override;
+
+    void consume(Chunk chunk) override;
+
+    void onFinish() override;
+
+    String getName() const override { return "EmbeddedRocksDBBulkSink"; }
+
+private:
+    /// Get a unique path to write temporary SST file
+    String getTemporarySSTFilePath();
+
+    /// Squash chunks to a minimum size
+    std::vector<Chunk> squash(Chunk chunk);
```

Comment:
The squashing logic here is different from `SquashingChunksTransform` logic. We don't create a new big chunk but just group some chunks together into a vector, which is much cheaper. `serializeChunks` will build final key-value columns from those chunks.

Reply:
Squashing logic here is different from `SquashingChunksTransform`. We don't create a new big chunk but grouping chunks in to a vector until we have enough rows, which is much cheaper. We can do this because `serializeChunks` can create key-value columns from multiple chunks.